### PR TITLE
feat: evm permit

### PR DESCRIPTION
**Pull request by enthusiastmartin**

### CODE REVIEW

File: rococo-local/config-zombienet.json [hunk 1]
```diff
@@ -40,14 +40,14 @@
       "collators": [
         {
           "name": "alice",
-          "command": "../target/release/hydradx",
+          "command": "/home/martin/workspace/galactic/HydraDX-node/target/release/hydradx",
```
Comment (mrq1911) [inline rococo-local/config-zombienet.json:43]: these changes are not intended probably

File: runtime/hydradx/src/evm/permit.rs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+		deadline: U256,
+		v: u8,
+		r: H256,
+		s: H256,
+	) -> DispatchResult {
+		let account_id = <R as pallet_evm::Config>::AddressMapping::into_account_id(source);
+		let account_nonce = frame_system::Pallet::<R>::account_nonce(&account_id);
+
+		let permit = pallet_evm_precompile_call_permit::CallPermitPrecompile::<R>::generate_permit(
+			precompiles::DISPATCH_ADDR,
```
Comment (mrq1911) [inline runtime/hydradx/src/evm/permit.rs:43]: we should support permit calls for other contracts as well

File: pallets/transaction-multi-payment/src/lib.rs [hunk 3]
```diff
@@ -341,6 +355,139 @@ pub mod pallet {
...
+			AccountCurrencyMap::<T>::insert(account_id.clone(), currency);
+			let result = T::EvmPermit::dispatch_permit(
+				source,
+				target,
+				input,
+				value,
+				gas_limit,
+				gas_price,
+				None,
+				None,
```
Comment (mrq1911) [inline pallets/transaction-multi-payment/src/lib.rs:400]: why is nonce handled differently?

File: pallets/transaction-multi-payment/src/lib.rs [hunk 4]
```diff
@@ -341,6 +355,139 @@ pub mod pallet {
...
 		}
+
+		// Dispatch EVM permit
+		#[pallet::call_index(4)]
+		#[pallet::weight(
+			<T as Config>::EvmPermit::dispatch_weight(*gas_limit)
+		)]
+		pub fn dispatch_permit(
+			origin: OriginFor<T>,
+			currency: AssetIdOf<T>,
```
Comment (mrq1911) [inline pallets/transaction-multi-payment/src/lib.rs:366]: thinking more about it this lets dispatcher to change payment currency without your approval. best would be to require user to call set_currency (though permit) 1st to change it same as we do with regular set_currency. regular evm call interaction with RPC (not permit) would always charge WETH. ofc the evm calls incoming from extrinsics would behave the same as they cuirrenctly are to enable substrate users to pay in any currency

File: pallets/transaction-multi-payment/src/lib.rs [hunk 5]
```diff
@@ -341,6 +355,139 @@ pub mod pallet {
 
 			Ok(())
 		}
+
+		// Dispatch EVM permit
+		#[pallet::call_index(4)]
+		#[pallet::weight(
+			<T as Config>::EvmPermit::dispatch_weight(*gas_limit)
+		)]
+		pub fn dispatch_permit(
```
Comment (mrq1911) [inline pallets/transaction-multi-payment/src/lib.rs:364]: where is nonce?

File: runtime/hydradx/src/evm/permit.rs [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+		+ pallet_evm::Config
+		+ pallet_transaction_multi_payment::Config
+		+ pallet_dynamic_evm_fee::Config,
+	R::Nonce: Into<U256>,
+	AccountId: From<R::AccountId>,
+{
+	fn validate_permit(
+		source: H160,
+		target: H160,
+		input: Vec<u8>,
```
Comment (mrq1911) [inline runtime/hydradx/src/evm/permit.rs:31]: this param should be called 'data' for consistency

**Review by mrq1911 — CHANGES_REQUESTED** [review]

File: precompiles/call-permit/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+		let permit_content = keccak_256(&permit_content);
+		let mut pre_digest = Vec::with_capacity(2 + 32 + 32);
+		pre_digest.extend_from_slice(b"\x19\x01");
+		pre_digest.extend_from_slice(&domain_separator);
+		pre_digest.extend_from_slice(&permit_content);
+		keccak_256(&pre_digest)
+	}
+
+	pub fn dispatch_inherent_cost() -> u64 {
+		3_000 // cost of ECRecover precompile for reference
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply precompiles/call-permit/src/lib.rs:127]: this comes from moonbeam, i dont want tochange anything in there

File: precompiles/call-permit/src/tests.rs [hunk 8]
```diff
@@ -0,0 +1,674 @@
...
+
+			precompiles()
+				.prepare_test(
+					CryptoAlith,
+					CallPermit,
+					PCall::nonces {
+						owner: Address(CryptoAlith.into()),
+					},
+				)
+				.expect_cost(0) // TODO: Test db read/write costs
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply precompiles/call-permit/src/tests.rs:189]: their tests

File: precompiles/call-permit/src/tests.rs [hunk 9]
```diff
@@ -0,0 +1,674 @@
...
+					},
+				)
+				.with_subcall_handle(move |_| panic!("should not perform subcall"))
+				.with_target_gas(Some(100_000 + dispatch_cost()))
+				.expect_cost(dispatch_cost())
+				.execute_reverts(|x| x == b"Call require too much gas (uint64 overflow)");
+		})
+}
+
+// // This test checks the validity of a metamask signed message against the permit precompile
```
[the remark this answers is not in the document]
Comment (enthusiastmartin) [reply precompiles/call-permit/src/tests.rs:448]: not other code

File: runtime/hydradx/src/system.rs [hunk 10]
```diff
@@ -487,6 +487,9 @@ impl pallet_transaction_multi_payment::Config for Runtime {
 	type EvmAssetId = evm::WethAssetId;
 	type InspectEvmAccounts = EVMAccounts;
 	type WeightToFee = WeightToFee;
+	type EvmPermit = evm::permit::EvmPermitHandler<Runtime>;
+	type TryCallCurrency<'a> = pallet_transaction_multi_payment::TryCallCurrency<Runtime>;
+	//type TryCallCurrency = pallet_transaction_multi_payment::TryCallCurrency<Runtime>;
```
Comment (mrq1911) [inline runtime/hydradx/src/system.rs:492]: delete

**Review by mrq1911 — CHANGES_REQUESTED** [review]
happened handler implementation in transaction multipayment should set default currency same for evm accounts

**Review by mrq1911 — APPROVED** [review]
ready for testing and UI impl

for production we need https://github.com/galacticcouncil/HydraDX-node/issues/814

**Pull request merged**